ambry-router/src/main/java/com.github.ambry.router/GetOperation.java

```diff
@@ -0,0 +1,204 @@
+    return operationCompleted;
+  }
+
+  /**
+   * Return the {@link MessageFormatFlags} to associate with the requests for this operation.
+   * @return the {@link MessageFormatFlags} to associate with the requests for this operation.
+   */
+  abstract MessageFormatFlags getOperationFlag();
+
+  /**
+   * For this operation, create and populate get requests to send out.
+   * @param requestRegistrationCallback the {@link RequestRegistrationCallback} to call for every request that gets
+   *                                    created as part of this poll operation.
+   */
+  abstract void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback);
+
+  /**
+   * Handle the given {@link ResponseInfo} received for a request that was sent out.
+   * @param responseInfo the {@link ResponseInfo} to be handled.
+   */
+  abstract void handleResponse(ResponseInfo responseInfo);
+
+  /**
+   * Process the given {@link ServerErrorCode} and set operation status accordingly.
+   * @param errorCode the {@link ServerErrorCode} to process.
+   */
+  void processServerError(ServerErrorCode errorCode) {
+    switch (errorCode) {
+      case Blob_Deleted:
+        logger.trace("Requested blob was deleted");
+        setOperationException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobDeleted));
+        break;
+      case Blob_Expired:
+        logger.trace("Requested blob is expired");
+        setOperationException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobExpired));
+        break;
+      case Blob_Not_Found:
+        logger.trace("Requested blob was not found on this server");
+        setOperationException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobDoesNotExist));
+        break;
+      default:
+        setOperationException(new RouterException("Server returned: " + errorCode, RouterErrorCode.AmbryUnavailable));
+        break;
+    }
+  }
+
+  /**
+   * Set the exception associated with this operation.
+   * A Blob_Deleted or Blob_Expired error overrides any other error previously received exception.
+   * @param exception the RouterException to possibly set.
+   */
+  void setOperationException(RouterException exception) {
+    if (operationException == null || exception.getErrorCode() == RouterErrorCode.BlobDeleted
+        || exception.getErrorCode() == RouterErrorCode.BlobExpired) {
+      operationException = exception;
+    }
+  }
+
+  /**
+   * Create and return the {@link GetRequest} associated with the given blobId.
+   * @return the crated {@link GetRequest}.
+   * @param blobId The {@link BlobId} for which the GetRequest is being created.
+   * @param flag The {@link MessageFormatFlags} to be set with the GetRequest.
+   * @return the created GetRequest.
+   */
+  protected GetRequest createGetRequest(BlobId blobId, MessageFormatFlags flag) {
+    ArrayList<BlobId> blobIds = new ArrayList<BlobId>(1);
+    blobIds.add(blobId);
```

Comment:
That is cool, thanks for showing me. Had to make a change to `PartitionRequestInfo` class to take in a `List` instead of an `ArrayList`.